contracts/crowdsale/crowdsale-mintable.aes

```diff
@@ -0,0 +1,165 @@
+contract ERC20 =
+  public function totalSupply : () => int
+  public function cap : () => int
+  public function balanceOf : address => int
+  public function allowance : (address, address) => int
+  public function transfer : (address, int) => bool
+  public function approve : (address, int) => bool
+  public function transferFrom : (address, address, int) => bool
+  public function increaseAllowance : (address, int) => bool
+  public function decreaseAllowance : (address, int) => bool
+  public function mint : (address, int) => bool
+  public function transferOwnership : (address) => () 
+  public function burn : int => bool
+
+
+contract Crowdsale =
+
+    record state = { 
+        erc20token : ERC20,
+        wallet : address, // Address where funds are collected
+        rate : int, 
+        endDate : int,
+        owner : address,
+        aeRaised : int}
+
+    public stateful function init(_erc20 : ERC20, _wallet : address, _rate : int, seconds : int) = 
+        require(_rate > 0, "")
+        require(_wallet != #0, "")
+        // require(_erc20 != #0, "") ??? what is default type
+        { erc20token = _erc20,
+          wallet = _wallet,
+          rate = _rate,
+          endDate = Chain.timestamp + seconds,
+          owner = Call.caller,
+          aeRaised = 0}
+
+    public function get() : int =
+        Chain.timestamp
+
+    // the address where funds are collected.
+    public function getWallet() : address =
+        state.wallet
+
+    // the number of token units a buyer gets per wei.
+    public function getRate() : int =
+        state.rate
+
+    // the amount of token raised.
+    public function aeRaised() : int =
+        state.aeRaised
+
+    // Consider calling buyTokens directly when purchasing tokens from a contract.
+    // _beneficiary: Recipient of the token purchase
+    public stateful function buyTokens(_beneficiary : address) =
+        isCrowdsale()
+        let _amount = Call.value
+
+        _preValidatePurchase(_beneficiary, _amount)
+        // calculate token amount to be created
+        let tokens = _getTokenAmount(_amount, state.rate)
+        
+        // update state
+        put(state{aeRaised = add(state.aeRaised, _amount)})
+
+        _processPurchase(_beneficiary, tokens)
+        // TODO: throw event something like TokensPurchased(Call.caller, _beneficiary, aeAmount, tokens);
+
+        _updatePurchasingState(_beneficiary, _amount)
+
+        _forwardFunds(_amount)
+        _postValidatePurchase(_beneficiary, _amount)
+
+    public function endCrowdsale() =
+        isOwner()
```

Comment:
it should be `isNotCrowdsale()` otherwise the owner might end the crowdsale beforehand